// @tag/tag-components/dist/collection/components/tag-http/tag-http.utils.js
import { findArrayInObject } from '../../shared/utils';
/**
 * Maps response data into IPagedData
 * @param {any[]} data - The response body data
 */
export const responseCollectionMapper = (response) => {
    const searchResponse = {
        data: getResponseData(response),
        error: false,
        page: getResponsePage(response),
        pageCount: getResponsePageCount(response),
        pageSize: getResponsePageSize(response),
        recordCount: getResponseTotalCount(response),
    };
    return searchResponse;
};
/**
 * Finds an array of data in a response and returns it
 * @param {any} data - The response body data
 */
const getResponseData = (response) => {
    const body = response && response.body;
    if (body === undefined) {
        return undefined;
    }
    else if (Array.isArray(body)) {
        return body;
    }
    else if (body.data && Array.isArray(body.data)) {
        return body.data;
    }
    else if (body.results && Array.isArray(body.results)) {
        return body.results;
    }
    else {
        return findArrayInObject(body);
    }
};
const getResponsePage = (response) => {
    const body = response && response.body;
    if (!body) {
        return undefined;
    }
    else if (body.page) {
        return body.page;
    }
    else if (body.pageNumber) {
        return body.pageNumber;
    }
    else {
        return undefined;
    }
};
const getResponsePageSize = (response) => {
    const body = response && response.body;
    if (!body) {
        return undefined;
    }
    else if (body.pageSize) {
        return body.pageSize;
    }
    else {
        return undefined;
    }
};
const getResponsePageCount = (response) => {
    const body = response && response.body;
    if (!body) {
        return undefined;
    }
    else if (body.pageCount) {
        return body.pageCount;
    }
    else if (body.pages) {
        return body.pages;
    }
    else {
        return undefined;
    }
};
const getResponseTotalCount = (response) => {
    const body = response && response.body;
    if (!body) {
        return undefined;
    }
    else if (body.count) {
        return body.count;
    }
    else if (body.totalCount) {
        return body.totalCount;
    }
    else if (body.recordCount) {
        return body.recordCount;
    }
    else {
        return undefined;
    }
};
